### ISSUE TITLE
poster_exporter_importer.py: Allow only exporting/importing posters

### ISSUE THREAD

**Issue by gitttt54** [Casvt/Plex-scripts#74]
Hello :-D

This is about music librarys and album art shown in Plex Clients. (maybe Movie/TV libraries can adopt this request later too)

**The problem:**
Short: Very often, when adding new music to Plex, the wrong cover-art is added to an album.

Long: As we know (me, at least, and a lot of Users at Plex Forum) Plex sucks at adding album art for music albums (embedded in the files or locally stored inside the album folder as cover.jpg). 
If a mp3 file has frontcover _and_ backcover embedded, PMS would chose the backcover in 40% instead of the frontcover (which mostly is preferred by the user). 
Also, if a cover.jpg is included in the folder, where the mp3 files resist, PMS still adds the backcover (or even disc-artwork, if embedded).
If so, the user has to correct the album art manually, which is a lot of work for large libraries.

**The solution:**
A script which checks each album folder for a cover.jpg and adds/changes the album art to that cover.jpg

**The result:**
Locally stored cover.jpg is used as album art for each album, instead of PMS chosen (wrong) album art.

Thank you for reading this and thinking about it :-D

P.S.
Actually, this SHOULD be done by PMS itself, but it is not. A lot of threads exist in the Plex forum, but unfortunately, they did not fix the bugs...

**Comment by Casvt** [comment]
So just basically: "Go through the folders of the plex library and upload poster images to the corresponding plex entry". So the opposite of [poster_downloader.py](https://github.com/Casvt/Plex-scripts/blob/main/various/poster_downloader.py), which downloads posters and puts it into files. You want a script that grabs files and uploads them. Then I'll add this request as a feature to that script (and also immediately music lib support). It's easily doable btw so don't worry about that.

**Comment by gitttt54** [comment]
You sumarized it very accurate. Thank you! 
I'm glad there is already a codebase you can tie in :-)

**Comment by Casvt** [comment]
Hey just had a thought. I proposed to integrate this request into poster_downloader.py. However, I can also integrate both poster_downloader.py together with this feature request into plex_exporter_importer.py. That last script can already export and import posters, only can it not (yet) _only_ do posters but that's easily fixable. And music library support needs to be added but that is already requested in issue #21.

So to summarise: I will add an argument to plex_exporter_importer.py to only do posters and will fulfil issue #21 (music lib support) which all together will fulfil your request.

**Comment by Casvt** [comment]
For the record (note to self): delete poster_downloader.py when this request is fulfilled.

**Comment by Casvt** [comment]
Added extra arguments to plex_exporter_importer.py that allow to more precisely select what to export/import. With these new args, you can now export only posters, or only metadata, or both, etc. etc. This means that when issue #21 is resolved, your request is fulfilled. 

For the record, this is the commit that added these new arguments: https://github.com/Casvt/Plex-scripts/commit/0f594062c88b00db63084d55742b11fa80e8455b

poster_downloader.py will be deleted as it is now replaced by plex_exporter_importer.py, where the user needs to setup args to only export posters. And as plex_exporter_importer.py also allows _importing_ posters, the functionality has even expanded.

**Comment by gitttt54** [comment]
Hi again :-)
I know, it might be unconventional, but would u mind to tell me/us/other interested people, what you did/programmed/changed? for people, who can not read or understand the code you wrote ;-)
I'm curious what barriers you took or how you managed to get this thing working.
(if you have spare time, of course! It's just some explaining, for us, who have no clue how hard programming is, or how easy some things can be accomplished...)

**Comment by Casvt** [comment]
If you run the script with `--NoWatched` and `--NoMetadata`, it will download only the posters of Plex media that is selected. Run the script with `-h` and all the arguments are explained.

Keep in mind that music libraries aren't supported yet so your request isn't fulfilled yet. And I'm planning on re-doing the arguments. Now you select what _not_ to export/import, but I'm going to change that to selecting what _to_ export/import. So you could just do `--Target posters` for example. Still planning on doing that.

A little explanation:
This script could already download and upload posters, only could it not _only_ do that. So I added extra arguments (`--NoMetadata`) that allow you to reduce the exporting/importing to only posters. When you don't export/import metadata (`--NoMetadata`) and watched status (`--NoWatched`), the only thing left over is posters.

But that's cheating. What you actually mean is, how do I download and upload the posters correctly? Well, I go to the selected media ("all shows in a library", "movie1") and for every entry, I grab the URL to the poster image. Then I connect to the link which downloads the poster and the result is put inside a file with .jpg extension. Tada, poster of media downloaded inside a file. We grab the folder that the media is in and put the poster file inside it so that it's alongside the media file.

When importing, we look at the folder that the Plex media is in and check if an accompanying poster file is in the folder too. If there is, we open it and upload the content to Plex saying "[this] is the picture that should be applied to [this] media"

**Comment by gitttt54** [comment]
Thank you very very much!

**Comment by Casvt** [comment]
> And the Music Library differs a lot from the Movie Library, so its not easy/simple to adapt?

Correct. Music libraries are totally different from movie and show libraries. So I need to write a lot of new code inside the file to support it, like: allowing you to select media in it (all, album, track), exporting the correct information, finding out the correct location to put the file in, the importing of all the information to the right place in Plex, etc. etc..

**Comment by Casvt** [comment]
Hey there! I've upgraded [plex_exporter_importer.py](https://github.com/Casvt/Plex-scripts/blob/main/multiple_servers/plex_exporter_importer.py) a lot today (spend 5 hours in total on it haha). The argument parsing has been improved a lot. You now select what you want to import/export instead of selecting what _not_ to import/export. This means that you can now just do `-t export --AllMusic -p poster` and kachow, all artist and album posters exported. Do `-t import` instead of `-t export` and all images are imported and applied (requires filenames to be `artist_thumb.jpg` and `album_thumb.jpg`).

And, as you can see in the arguments I showed, there also is now music library support! Basically what I want to say is that the script has been improved a lot and that it works amazing. It also now supports what you requested: you can upload (and auto apply) artist and album images to plex. It does require that the filenames of the images are `artist_thumb.jpg` and `album_thumb.jpg` but that's a simple search and rename for you to do. If you don't have an image for something, don't worry, the script will just skip over it then.

Enjoy! And feel free to ask any question when needed ;)

For the record, this commit added music library support: https://github.com/Casvt/Plex-scripts/commit/2869bd85920a2f84f8593c0f4bb82737f0bce2dd
For the record, this commit upgraded the argument parsing: https://github.com/Casvt/Plex-scripts/commit/991e67b30a5a94d56114b75d196d3e88b3da1219

**Comment by gitttt54** [comment]
Wow, thank you so much for this, it reads great and I'm eager to try it out....but - unfortunately - I need to ask another question before :-)

Can you add an option to define the artist and album image name?
Like `--AlbumImageName cover_front.jpg` and `--ArtistImageName Artist.jpg`
So users can define their own filenames to be imported.

Its simply because:
I can not rename all the images, because other software scripts won't work then.
And I share (better, they share with me) folders with friends, which surely have their own scripts and would not like me to rename all their image files.

That would be great and very usefull too and it will hopefully not take a lot of time for you, because you investigated so much time (I think its very fast anyways...)

Thank you!

**Comment by Casvt** [comment]
Yeah. That's the thing. The idea of the script is that you export the library (so the script creates jpg's) and then you basically have a backup. You can start a whole new Plex server, import the library, run the script in import mode and everything is back. So there was no need to support alternate file naming as that was not the idea for how this script was supposed to be used.

But I get what you mean. If I integrate this feature request, people can import their poster files, and then export again and then they have their poster files back but correctly named. Or only run the script in import mode to just apply any external poster files (which is your use case).

I think that its relevant enough to integrate it yeah. And with relevant, I mean if it is useful for enough people etc. etc. And I think it is.

Your request is completed here. The script can now do what you requested. Only do you have a special situation where the filenames don't match up. That's a new "case". It's important for you, but not relavant to this specific request here. So this is what we're going to do: please make a new feature request where you request alternate filenames to be searched for when importing. And then we'll go on from there alright? :)

**Comment by gitttt54** [comment]
> Your request is completed here. [...] please make a new feature request [...] 

The new request is [here](https://github.com/Casvt/Plex-scripts/issues/107). Thank you so much for taking the time and think about my ideas/needs ;-) Have a good time!
